### ISSUE TITLE
Heredoc warnings SC1009, SC1073, SC1041, SC1042

### ISSUE THREAD

**Issue by joaquinabian** [koalaman/shellcheck#2103]
#### For bugs
vscode-shellcheck v0.12.1
win10 64 bits

#### Here's a snippet or screenshot that shows the problem:

```
message=$(cat << EOF
Is this raspberry pi a
   - Master (.211)
   - Slave  (.212)
   - Test   (.215)
   ¿ M / S / T ?  default T -> 
EOF
)
```

#### Here's what shellcheck currently says:
On first line:
SC1009 The mentioned parser error was in this simple command
SC1073 Couldn't parse this here document. Fix to allow more checks.
SC1041 Found 'eof' further down, but not on a separate line.
SC1042 -

#### Here's what I wanted or expected to see:
Nothing ?

**Comment by brother** [comment]
What shell type is in play here? 
And shellcheck version is omitted as well.

https://www.shellcheck.net reports no errors for neither sh nor bash.

**Comment by joaquinabian** [comment]
/bin/bash
Not sure about shellcheck version. The vscode-shellcheck plugin of Visual Studio Code uses a precompiled shellcheck binary already bundled with it. I cannot find how to get the version.
I agree there are no warnings coming from shellcheck.net.

**Comment by brother** [comment]
The shellcheck version is available in the terminal area of vs code. I get 0.7.1 on my linux box (same as at my system, close to git HEAD).
I have no errors reported using this extension in vs code, as was demonstrated via the online version.

![image](https://user-images.githubusercontent.com/243351/101832970-993c3800-3b38-11eb-9dfa-8e852891d308.png)

Either the example is not complete and hides other errors or the shellcheck version is older. The problem that is pointed out in the screenshot is that of `message` not being used.

![image](https://user-images.githubusercontent.com/243351/101833061-b7099d00-3b38-11eb-969c-6d9e0c3400dc.png)

**Comment by joaquinabian** [comment]
Caught the problem.
Warnings were a bit misleading. They were not related with the piece of code but with line endings.
I code the same application with VSCode sometimes in windows, sometimes in ubuntu, push/pulling from Github.
I saw the warnings today again in windows. 
Line endings in VSCode indicated CRLF. When changed to LF, all four warnings disappeared.

**Comment by brother** [comment]
git can handle the lineendings for you if you want =)
https://docs.github.com/en/free-pro-team@latest/github/using-git/configuring-git-to-handle-line-endings

**Comment by joaquinabian** [comment]
Thanks Martin,
This was actually my problem.
I have `git config core.autocrlf` to change line endings depending on the OS I am pulling.  

In Windows, my git-pulled files get the CLRF ending as expected, but then I got a bunch of spellcheck errors (SC1017).
I can use VSCode configuration to eliminate them: `"shellcheck.exclude": ["1017"]` but this way I still have the heredoc warnings errors I reported in this ticket.

The solution I tried initially was to change manually CLRF to LF  in VSCode for every file (but this is detected by git as a change to commit. No convenient).

I read again, more carefully, the link you sent and it seems the solution is to set git configuration per repository.

Still, there is the issue (maybe not really a issue ?) that, in Windows, with Windows line-endings, Spellcheck gives so many line-end SC1017 warnings (this can be controlled though) and cannot recognize these line-endings as such when parsing heredocs (I think this is what SC1041 shows).

**Comment by brother** [comment]
The last thing sounds like a real problem indeed. Maybe the title and body of this issue should be updated with the findings

**Comment by koalaman** [comment]
This was a surprisingly interesting set of bugs: 1. The shebang should have given a carriage return warning, but it was being delayed and eventually discarded, 2. Bash allows carriage returns to be part of the token, so this code is actually valid with the carriage returns. ShellCheck did not support this. 3. The SC1042 error should have given a helpful `Close matches include 'EOF\r' (!= 'EOF').` but it failed to escape the carriage return, so the message was unreadable.

All these issues have now been fixed. Thanks!
